packages/sdk-core/src/proving/proving-manager-thread.ts

```diff
@@ -98,38 +124,94 @@ export class ProvingManagerWrapper {
-    }
-
-    if (pmSetupInput.refreshCommitment) {
-      pm.setRefreshCommitment(pmSetupInput.refreshCommitment);
+    const pm = new Manager(protocol);
+
+    if (protocol === 'mixer') {
+      const input = pmSetupInput as MixerPMSetupInput;
+      const { note } = await Note.deserialize(input.note);
+
+      pm.setLeaves(input.leaves);
+      pm.setRelayer(input.relayer);
+      pm.setRecipient(input.recipient);
+      pm.setLeafIndex(String(input.leafIndex));
+      pm.setRefund(String(input.refund));
+      pm.setFee(String(input.fee));
+      pm.setPk(u8aToHex(input.provingKey).replace('0x', ''));
+      pm.setNote(note);
+    } else if (protocol === 'anchor') {
+      const input = pmSetupInput as AnchorPMSetupInput;
+      const { note } = await Note.deserialize(input.note);
+
+      pm.setLeaves(input.leaves);
+      pm.setRelayer(input.relayer);
+      pm.setRecipient(input.recipient);
+      pm.setLeafIndex(String(input.leafIndex));
+      pm.setRefund(String(input.refund));
+      pm.setFee(String(input.fee));
+      pm.setPk(u8aToHex(input.provingKey).replace('0x', ''));
+      pm.setNote(note);
+      pm.setRoots(input.roots);
+      pm.setRefreshCommitment(input.refreshCommitment);
+    } else if (protocol === 'vanchor') {
+      const input = pmSetupInput as VAnchorPMSetupInput;
+      const metaDataNote = input.metaDataNote || input.inputNotes[0];
+      const { note } = await Note.deserialize(metaDataNote);
+      const notes = await Promise.all(input.inputNotes.map((note) => Note.deserialize(note)));
+      const jsNotes = notes.map((note) => note.note);
+
+      pm.setNote(note);
+      pm.setNotes(jsNotes);
+      pm.setIndices(input.indices.map((i) => i.toString()) as any);
+      pm.setPk(u8aToHex(input.provingKey).replace('0x', ''));
+      pm.setRoots(input.roots);
+      pm.chain_id(input.chainId);
+      pm.public_amount(input.publicAmount);
+      pm.setVanchorOutputConfig(...input.outputConfigs);
+      pm.setExtDatahash(input.externalDataHash);
+      // leaves insertion
+      const wasm = await this.wasmBlob;
+      const leavesMap = new wasm.LeavesMapInput();
+
+      for (const key of Object.keys(input.leavesMap)) {
+        leavesMap.setChainLeaves(key as any, input.leavesMap[key]);
+      }
+
+      pm.setLeavesMap(leavesMap);
+    } else {
+      throw new Error('invalid protocol');
     }
 
     const proofInput = pm.build_js();
-    const proof = await this.generateProof(proofInput);
-
-    return {
-      nullifierHash: proof.nullifierHash,
-      proof: proof.proof,
-      root: proof.root,
-      roots: proof.roots
-    };
+
+    const proofOutput = await this.generateProof(proofInput);
+
+    if (protocol === 'mixer' || protocol === 'anchor') {
+      const proof = proofOutput.proof;
+
+      const proofPayload: ProofI<'anchor'> = {
+        nullifierHash: proof.nullifierHash,
+        proof: proof.proof,
+        root: proof.root,
+        roots: proof.roots
+      };
```

Comment:
Should the mixer return a different `ProofI<>`?

Reply:
mixer/anchor shares the same proof type, but here it should be `mixer` for sure just for being semantic

Reply:
We should separate this as mixer doesn't have roots object.

Reply:
@AhmedKorim do you agree?

Reply:
This is a wapper around what `arkworks-gadgets` exposes
```rust

pub struct MixerProof {
	pub proof: Vec<u8>,
	pub leaf_raw: Vec<u8>,
	pub nullifier_hash_raw: Vec<u8>,
	pub root_raw: Vec<u8>,
	pub public_inputs_raw: Vec<Vec<u8>>,
}

```

we can do that update after doing it in  `arkworks-gadgets`

Reply:
https://github.com/webb-tools/arkworks-gadgets/blob/a2c494c30d3ee31610f1eb6576c8e3acd3c0cf1a/arkworks-setups/src/common.rs#L54-L54

Reply:
you are right guys